File: _reimplementations/Python/tvtools/tests/test_tvevent.py
"""
Comprehensive tests for tvevent Python implementation
"""

import pytest
import pandas as pd
import numpy as np
import sys
import os

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from tvtools import tvexpose, tvevent

DATA_PATH = "/home/ubuntu/Stata-Tools/_testing/data"


@pytest.fixture
def test_data():
    """Load test data."""
    import pyreadstat
    cohort, _ = pyreadstat.read_dta(f"{DATA_PATH}/cohort.dta")
    hrt, _ = pyreadstat.read_dta(f"{DATA_PATH}/hrt.dta")
    return {'cohort': cohort, 'hrt': hrt}


@pytest.fixture
def intervals_data(test_data):
    """Create interval data for tvevent testing."""
    result = tvexpose(
        master_data=test_data['cohort'],
        exposure_file=test_data['hrt'],
        id='id', start='rx_start', stop='rx_stop', exposure='hrt_type',
        entry='study_entry', exit='study_exit', reference=0,
        generate='hrt_exp', verbose=False
    )
    return result.data


@pytest.fixture
def events_data(test_data):
    """Create events data for testing."""
    events = test_data['cohort'][['id', 'edss4_dt', 'death_dt', 'emigration_dt']].copy()
    events = events.rename(columns={'edss4_dt': 'event_date'})
    return events


class TestTVEventBasic:
    """Basic functionality tests."""

    def test_basic_single_event(self, intervals_data, events_data):
        """Test basic single event creation."""
        result = tvevent(
            intervals_data=intervals_data,
            events_data=events_data,
            id='id',
            date='event_date',
            generate='outcome',
            type='single'
        )

        assert result is not None
        assert len(result.data) > 0
        assert 'outcome' in result.data.columns
        assert result.N_events >= 0

    def test_competing_risks(self, intervals_data, events_data):
        """Test competing risks handling."""
        result = tvevent(
            intervals_data=intervals_data,
            events_data=events_data,
            id='id',
            date='event_date',
            compete=['death_dt', 'emigration_dt'],
            generate='outcome',
            type='single'
        )

        # Should have multiple event types
        outcome_vals = result.data['outcome'].unique()
        assert len(outcome_vals) > 1


class TestTVEventTimeGeneration:
    """Time variable generation tests."""

    def test_timegen_days(self, intervals_data, events_data):
        """Test time generation in days."""
        result = tvevent(
            intervals_data=intervals_data,
            events_data=events_data,
            id='id',
            date='event_date',
            generate='outcome',
            timegen='followup',
            timeunit='days',
            type='single'
        )

        assert 'followup' in result.data.columns
        assert all(result.data['followup'] >= 0)

    def test_timegen_months(self, intervals_data, events_data):
        """Test time generation in months."""
        result = tvevent(
            intervals_data=intervals_data,
            events_data=events_data,
            id='id',
            date='event_date',
            generate='outcome',
            timegen='followup',
            timeunit='months',
            type='single'
        )

        assert 'followup' in result.data.columns

    def test_timegen_years(self, intervals_data, events_data):
        """Test time generation in years."""
        result = tvevent(
            intervals_data=intervals_data,
            events_data=events_data,
            id='id',
            date='event_date',
            generate='outcome',
            timegen='followup',
            timeunit='years',
            type='single'
        )

        assert 'followup' in result.data.columns


class TestTVEventTypes:
    """Event type handling tests."""

    def test_single_event_type(self, intervals_data, events_data):
        """Test single event type censors after first event."""
        result = tvevent(
            intervals_data=intervals_data,
            events_data=events_data,
            id='id',
            date='event_date',
            generate='outcome',
            type='single'
        )

        # For each person, max 1 event
        events_per_person = result.data[result.data['outcome'] > 0].groupby('id').size()
        assert all(events_per_person <= 1)

    def test_recurring_events(self):
        """Test recurring events type."""
        intervals = pd.DataFrame({
            'id': [1, 1, 1, 2, 2],
            'start': [1, 11, 21, 1, 11],
            'stop': [10, 20, 30, 10, 20],
            'exposure': [1, 1, 0, 1, 0]
        })

        events = pd.DataFrame({
            'id': [1, 1, 2],
            'event_date': [5.0, 15.0, 8.0]
        })

        result = tvevent(
            intervals_data=intervals,
            events_data=events,
            id='id',
            date='event_date',
            generate='outcome',
            type='recurring'
        )

        assert len(result.data) > 0


class TestTVEventValidation:
    """Validation tests."""

    def test_invalid_type_raises(self, intervals_data, events_data):
        """Test that invalid type raises error."""
        with pytest.raises(ValueError, match="single.*recurring"):
            tvevent(
                intervals_data=intervals_data,
                events_data=events_data,
                id='id',
                date='event_date',
                generate='outcome',
                type='invalid'
            )

    def test_invalid_timeunit_raises(self, intervals_data, events_data):
        """Test that invalid timeunit raises error."""
        with pytest.raises(ValueError, match="days.*months.*years"):
            tvevent(
                intervals_data=intervals_data,
                events_data=events_data,
                id='id',
                date='event_date',
                generate='outcome',
                timegen='time',
                timeunit='invalid',
                type='single'
            )

    def test_missing_columns_raises(self, intervals_data, events_data):
        """Test that missing columns raise error."""
        with pytest.raises(ValueError, match="not found"):
            tvevent(
                intervals_data=intervals_data,
                events_data=events_data,
                id='id',
                date='invalid_column',
                generate='outcome',
                type='single'
            )


class TestTVEventReturns:
    """Return value tests."""

    def test_returns_structure(self, intervals_data, events_data):
        """Test proper return structure."""
        result = tvevent(
            intervals_data=intervals_data,
            events_data=events_data,
            id='id',
            date='event_date',
            generate='outcome',
            type='single'
        )

        assert hasattr(result, 'data')
        assert hasattr(result, 'N')
        assert hasattr(result, 'N_events')
        assert hasattr(result, 'generate')
        assert hasattr(result, 'type')


if __name__ == '__main__':
    pytest.main([__file__, '-v'])

File: _reimplementations/R/tvtools/tests/testthat/test_tvevent.R
# Comprehensive tests for tvevent R implementation
# Based on Stata test suite from _testing/test_tvevent.do

library(testthat)
library(tvtools)
suppressMessages(library(dplyr))

# Test data path - use RDS files
data_path <- "/home/tpcopeland/Stata-Tools/_reimplementations/data/R"

# Helper function to load test data
load_test_data <- function() {
  cohort <- readRDS(file.path(data_path, "cohort.rds"))
  hrt <- readRDS(file.path(data_path, "hrt.rds"))
  list(cohort = cohort, hrt = hrt)
}

# Helper to create interval data for testing
create_intervals <- function(data) {
  result <- tvexpose(
    master_data = data$cohort,
    exposure_file = data$hrt,
    id = "id",
    start = "rx_start",
    stop = "rx_stop",
    exposure = "hrt_type",
    entry = "study_entry",
    exit = "study_exit",
    reference = 0,
    generate = "hrt_exp",
    verbose = FALSE
  )
  result$data
}

# Helper to create event data
create_events <- function(data) {
  events <- data$cohort[, c("id", "edss4_dt", "death_dt", "emigration_dt")]
  names(events)[2] <- "event_date"
  events
}

# ============================================================================
# TEST 1: Basic Single Event
# ============================================================================
test_that("tvevent creates basic single event flags", {
  data <- load_test_data()
  intervals <- create_intervals(data)
  events <- create_events(data)

  result <- tvevent(
    intervals_data = intervals,
    events_data = events,
    id = "id",
    date = "event_date",
    generate = "outcome",
    type = "single"
  )

  expect_true(is.list(result))
  expect_true("data" %in% names(result))
  expect_true(nrow(result$data) > 0)
  expect_true("outcome" %in% names(result$data))
  expect_true(result$N_events >= 0)
})

# ============================================================================
# TEST 2: Competing Risks
# ============================================================================
test_that("tvevent handles competing risks", {
  data <- load_test_data()
  intervals <- create_intervals(data)
  events <- create_events(data)

  result <- tvevent(
    intervals_data = intervals,
    events_data = events,
    id = "id",
    date = "event_date",
    compete = c("death_dt", "emigration_dt"),
    generate = "outcome",
    type = "single"
  )

  expect_true(is.list(result))
  # Should have multiple event types
  outcome_vals <- unique(result$data$outcome)
  expect_true(length(outcome_vals) > 1)
})

# ============================================================================
# TEST 3: Time Generation - Days
# ============================================================================
test_that("tvevent timegen option creates time variable in days", {
  data <- load_test_data()
  intervals <- create_intervals(data)
  events <- create_events(data)

  result <- tvevent(
    intervals_data = intervals,
    events_data = events,
    id = "id",
    date = "event_date",
    generate = "outcome",
    timegen = "followup_time",
    timeunit = "days",
    type = "single"
  )

  expect_true("followup_time" %in% names(result$data))
  # All times should be non-negative
  expect_true(all(result$data$followup_time >= 0))
})

# ============================================================================
# TEST 4: Time Generation - Months
# ============================================================================
test_that("tvevent timegen option creates time variable in months", {
  data <- load_test_data()
  intervals <- create_intervals(data)
  events <- create_events(data)

  result <- tvevent(
    intervals_data = intervals,
    events_data = events,
    id = "id",
    date = "event_date",
    generate = "outcome",
    timegen = "followup_time",
    timeunit = "months",
    type = "single"
  )

  expect_true("followup_time" %in% names(result$data))
})

# ============================================================================
# TEST 5: Time Generation - Years
# ============================================================================
test_that("tvevent timegen option creates time variable in years", {
  data <- load_test_data()
  intervals <- create_intervals(data)
  events <- create_events(data)

  result <- tvevent(
    intervals_data = intervals,
    events_data = events,
    id = "id",
    date = "event_date",
    generate = "outcome",
    timegen = "followup_time",
    timeunit = "years",
    type = "single"
  )

  expect_true("followup_time" %in% names(result$data))
})

# ============================================================================
# TEST 6: Recurring Events
# ============================================================================
test_that("tvevent handles recurring events", {
  # Create simple recurring event test data
  intervals <- data.frame(
    id = c(1, 1, 1, 2, 2),
    start = c(1, 11, 21, 1, 11),
    stop = c(10, 20, 30, 10, 20),
    exposure = c(1, 1, 0, 1, 0)
  )

  events <- data.frame(
    id = c(1, 1, 2),
    event_date = c(5, 15, 8)
  )

  result <- tvevent(
    intervals_data = intervals,
    events_data = events,
    id = "id",
    date = "event_date",
    generate = "outcome",
    type = "recurring"
  )

  expect_true(is.list(result))
  expect_true(nrow(result$data) > 0)
})

# ============================================================================
# TEST 7: Replace Option
# ============================================================================
test_that("tvevent replace option overwrites existing variable", {
  data <- load_test_data()
  intervals <- create_intervals(data)
  events <- create_events(data)

  # First create outcome
  intervals$outcome <- 0

  result <- tvevent(
    intervals_data = intervals,
    events_data = events,
    id = "id",
    date = "event_date",
    generate = "outcome",
    type = "single",
    replace = TRUE
  )

  expect_true("outcome" %in% names(result$data))
})

# ============================================================================
# TEST 8: Custom Event Labels
# ============================================================================
test_that("tvevent eventlabel option customizes labels", {
  data <- load_test_data()
  intervals <- create_intervals(data)
  events <- create_events(data)

  result <- tvevent(
    intervals_data = intervals,
    events_data = events,
    id = "id",
    date = "event_date",
    compete = c("death_dt", "emigration_dt"),
    generate = "outcome",
    eventlabel = c("0" = "No Event", "1" = "Primary", "2" = "Death", "3" = "Emigration"),
    type = "single"
  )

  # Should complete without error and have outcome variable
  expect_true("outcome" %in% names(result$data))
})

# ============================================================================
# TEST 9: Continuous Variable Adjustment
# ============================================================================
test_that("tvevent continuous option adjusts cumulative variables", {
  # Create intervals with cumulative dose
  intervals <- data.frame(
    id = c(1, 1, 2),
    start = c(1, 11, 1),
    stop = c(10, 20, 30),
    exposure = c(1, 1, 0),
    cum_dose = c(100, 200, 0)
  )

  # Event in middle of interval
  events <- data.frame(
    id = c(1),
    event_date = c(15)
  )

  result <- tvevent(
    intervals_data = intervals,
    events_data = events,
    id = "id",
    date = "event_date",
    generate = "outcome",
    continuous = c("cum_dose"),
    type = "single"
  )

  expect_true(is.list(result))
  expect_true("cum_dose" %in% names(result$data))
})

# ============================================================================
# TEST 10: KeepVars Option
# ============================================================================
test_that("tvevent keepvars option merges additional variables", {
  data <- load_test_data()
  intervals <- create_intervals(data)

  # Create events with extra variables
  events <- data$cohort[, c("id", "edss4_dt", "death_dt", "female", "age")]
  names(events)[2] <- "event_date"

  result <- tvevent(
    intervals_data = intervals,
    events_data = events,
    id = "id",
    date = "event_date",
    generate = "outcome",
    keepvars = c("female", "age"),
    type = "single"
  )

  # Note: keepvars from events are only merged to event intervals
  expect_true(is.list(result))
})

# ============================================================================
# TEST 11: Empty Events Data
# ============================================================================
test_that("tvevent handles empty events data", {
  data <- load_test_data()
  intervals <- create_intervals(data)

  # Create empty events
  events <- data.frame(
    id = integer(0),
    event_date = numeric(0)
  )

  # Should produce all censored
  expect_warning(
    result <- tvevent(
      intervals_data = intervals,
      events_data = events,
      id = "id",
      date = "event_date",
      generate = "outcome",
      type = "single"
    )
  )

  expect_equal(result$N_events, 0)
})

# ============================================================================
# TEST 12: All Missing Event Dates
# ============================================================================
test_that("tvevent handles all missing event dates", {
  data <- load_test_data()
  intervals <- create_intervals(data)

  # Create events with all NA dates (as numeric NA)
  events <- data.frame(
    id = c(1, 2, 3),
    event_date = as.numeric(c(NA, NA, NA))
  )

  expect_warning(
    result <- tvevent(
      intervals_data = intervals,
      events_data = events,
      id = "id",
      date = "event_date",
      generate = "outcome",
      type = "single"
    )
  )
})

# ============================================================================
# TEST 13: Single-Day Intervals
# ============================================================================
test_that("tvevent handles single-day intervals", {
  # Create intervals with single-day entries
  intervals <- data.frame(
    id = c(1, 1, 2),
    start = c(1, 10, 1),
    stop = c(10, 10, 20),  # Day 10 is single-day
    exposure = c(1, 0, 1)
  )

  events <- data.frame(
    id = c(1, 2),
    event_date = c(5, 15)
  )

  result <- tvevent(
    intervals_data = intervals,
    events_data = events,
    id = "id",
    date = "event_date",
    generate = "outcome",
    type = "single"
  )

  expect_true(nrow(result$data) > 0)
})

# ============================================================================
# TEST 14: Returns Structure
# ============================================================================
test_that("tvevent returns proper structure", {
  data <- load_test_data()
  intervals <- create_intervals(data)
  events <- create_events(data)

  result <- tvevent(
    intervals_data = intervals,
    events_data = events,
    id = "id",
    date = "event_date",
    generate = "outcome",
    type = "single"
  )

  expect_true("N" %in% names(result))
  expect_true("N_events" %in% names(result))
  expect_true("generate" %in% names(result))
  expect_true("type" %in% names(result))
})

# ============================================================================
# TEST 15: Class Attribute
# ============================================================================
test_that("tvevent returns object of correct class", {
  data <- load_test_data()
  intervals <- create_intervals(data)
  events <- create_events(data)

  result <- tvevent(
    intervals_data = intervals,
    events_data = events,
    id = "id",
    date = "event_date",
    generate = "outcome",
    type = "single"
  )

  expect_true("tvevent" %in% class(result))
})

# ============================================================================
# Run all tests
# ============================================================================
cat("\nRunning tvevent tests...\n")
